test/csuite/random_directio/main.c

```diff
@@ -0,0 +1,1127 @@
+check_one_entry(WT_SESSION *session, const char *uri, const char *key,
+    const char *value)
+{
+	WT_CURSOR *cursor;
+	int ret;
+	char *gotkey, *gotvalue;
+
+	testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+	testutil_check(cursor->next(cursor));
+	cursor->get_key(cursor, &gotkey);
+	cursor->get_value(cursor, &gotvalue);
+	testutil_assert(WT_STREQ(key, gotkey));
+	testutil_assert(WT_STREQ(value, gotvalue));
+	ret = cursor->next(cursor);
+	testutil_assert(ret == WT_NOTFOUND);
+	testutil_check(cursor->close(cursor));
+}
+
+/*
+ * check_schema
+ *	Check that the database has the expected schema according to the
+ *	last id seen for this thread.
+ */
+static void
+check_schema(WT_SESSION *session, uint64_t lastid, uint32_t threadid)
+{
+	char uri[50], uri2[50];
+
+	if (SCHEMA_OP(lastid, 0)) {
+		/* Create table operation. */
+		gen_table_name(uri, sizeof(uri), lastid, threadid);
+		check_empty(session, uri);
+	}
+	if (SCHEMA_OP(lastid, 1)) {
+		/* Insert value operation. */
+		gen_table_name(uri, sizeof(uri), lastid - 1, threadid);
+		check_one_entry(session, uri, uri, uri);
+	}
+	if (SCHEMA_OP(lastid, 2)) {
+		/* Table rename operation. */
+		gen_table_name(uri, sizeof(uri), lastid - 2, threadid);
+		gen_table2_name(uri2, sizeof(uri2), lastid - 2, threadid);
+		check_dropped(session, uri);
+		check_one_entry(session, uri2, uri, uri);
+	}
+	if (SCHEMA_OP(lastid, 3)) {
+		/* Value update operation. */
+		gen_table_name(uri, sizeof(uri), lastid - 2, threadid);
+		gen_table2_name(uri2, sizeof(uri2), lastid - 2, threadid);
+		check_one_entry(session, uri2, uri, uri2);
+	}
+	if (SCHEMA_OP(lastid, 4)) {
+		/* Drop table operation. */
+		gen_table2_name(uri2, sizeof(uri2), lastid - 2, threadid);
+		check_dropped(session, uri2);
+	}
+}
+
+/*
+ * check_db --
+ *	Make a copy of the database and verify its contents.
+ */
+static bool
+check_db(uint32_t nth, uint32_t datasize, bool directio, bool schema_test)
+{
+	struct sigaction sa;
+	WT_CONNECTION *conn;
+	WT_CURSOR *cursor, *meta, *rev;
+	WT_SESSION *session;
+	uint64_t gotid, id;
+	uint64_t *lastid;
+	uint32_t gotth, kvsize, th, threadmap;
+	int ret, status;
+	char buf[4096];
+	char *gotkey, *gotvalue, *keybuf, *p;
+	char **large_arr;
+
+	keybuf = dcalloc(datasize, 1);
+	lastid = dcalloc(nth, sizeof(uint64_t));
+
+	large_arr = dcalloc(nth, sizeof(char *));
+	for (th = 0; th < nth; th++) {
+		large_arr[th] = dcalloc(LARGE_WRITE_SIZE, 1);
+		large_buf(large_arr[th], LARGE_WRITE_SIZE, th, true);
+	}
+
+	/*
+	 * We make a copy of the directory (possibly using direct IO)
+	 * for recovery and checking, and an identical copy that
+	 * keeps the state of all files before recovery starts.
+	 */
+	testutil_check(__wt_snprintf(buf, sizeof(buf),
+	    "H='%s'; C=$H.CHECK; S=$H.SAVE; rm -rf $C $S;"
+	    " mkdir $C $S; for f in `ls $H/`; do "
+	    " dd if=$H/$f of=$C/$f bs=4096 %s >/dev/null 2>&1 || exit 1; done;"
+	    " cp -pr $C $S",
+	    home, directio ? "iflag=direct" : ""));
```

Comment:
It would be more helpful if `WT_TEST.random-directio.SAVE` ended up with the files in the top level directory rather than `WT_TEST.random-directio.SAVE/WT_TEST.random-directio.CHECK/<files>`.